pkg/nvlib/info/resolver.go

```diff
@@ -0,0 +1,64 @@
+/**
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package info
+
+// Platform represents a supported plaform.
+type Platform string
+
+const (
+	PlatformAuto    = Platform("auto")
+	PlatformNVML    = Platform("nvml")
+	PlatformTegra   = Platform("tegra")
+	PlatformWSL     = Platform("wsl")
+	PlatformUnknown = Platform("unknown")
+)
+
+type platformResolver struct {
+	logger   basicLogger
+	platform Platform
+	Properties
+}
+
+func (p platformResolver) ResolvePlatform() Platform {
```

Comment:
This abstraction seems nice and straightforward. Where will it be consumed exactly? Is it meant as a replacement for direct calls to `HasNvml()` `IsTegraSystem()` etc.?

Reply:
For most cases, yes, the `ResolvePlatform()` call will replace the individual calls and replicating the conditional logic.

I have already updated the `nvidia-container-toolkit` and `device-plugin` to make use of this functionality:
* https://github.com/NVIDIA/nvidia-container-toolkit/pull/428/files#diff-61ad274e8f1d9eb6d3aa8c961e3bf217f50363406924b4b4b1e40f049a5bb390R49
* https://github.com/NVIDIA/nvidia-container-toolkit/pull/428/files#diff-e43f2243963e5310d841effc7b372f8ecb0f27b06a0a9fd374fd3196101d2410R195
* https://github.com/NVIDIA/k8s-device-plugin/pull/673/files#diff-a9bc30e8fa1d6bf81ad74b60218fd217abfb27090d4f72e187e7f2a19ee4e82fR100
* https://github.com/NVIDIA/k8s-device-plugin/pull/673/files#diff-66f1319e3c95c17402a1547b5ae54f94e114de92111f21648ebe1fd635d6880dR45

Note that the "Properties" themselves may still be required for edge cases, but we could probably make them internal at somet point.